runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyStageFuser.java

```diff
@@ -0,0 +1,176 @@
+    Environment environment = getStageEnvironment(pipeline, initialNodes);
+
+    ImmutableSet.Builder<PTransformNode> fusedTransforms = ImmutableSet.builder();
+    fusedTransforms.addAll(initialNodes);
+
+    Set<PCollectionNode> fusedCollections = new LinkedHashSet<>();
+    Set<PCollectionNode> materializedPCollections = new LinkedHashSet<>();
+
+    Queue<PCollectionNode> fusionCandidates = new ArrayDeque<>();
+    for (PTransformNode initialConsumer : initialNodes) {
+      fusionCandidates.addAll(pipeline.getOutputPCollections(initialConsumer));
+    }
+    while (!fusionCandidates.isEmpty()) {
+      PCollectionNode candidate = fusionCandidates.poll();
+      if (fusedCollections.contains(candidate) || materializedPCollections.contains(candidate)) {
+        // This should generally mean we get to a Flatten via multiple paths through the graph and
+        // we've already determined what to do with the output.
+        LOG.debug(
+            "Skipping fusion candidate {} because it is {} in this {}",
+            candidate,
+            fusedCollections.contains(candidate) ? "fused" : "materialized",
+            ExecutableStage.class.getSimpleName());
+        continue;
+      }
+      PCollectionFusibility fusibility =
+          canFuse(pipeline, candidate, environment, fusedCollections);
+      switch (fusibility) {
+        case MATERIALIZE:
+          materializedPCollections.add(candidate);
+          break;
+        case FUSE:
+          // All of the consumers of the candidate PCollection can be fused into this stage. Do so.
+          fusedCollections.add(candidate);
+          fusedTransforms.addAll(pipeline.getPerElementConsumers(candidate));
+          for (PTransformNode consumer : pipeline.getPerElementConsumers(candidate)) {
+            // The outputs of every transform fused into this stage must be either materialized or
+            // themselves fused away, so add them to the set of candidates.
+            fusionCandidates.addAll(pipeline.getOutputPCollections(consumer));
+          }
+          break;
+        default:
+          throw new IllegalStateException(
+              String.format(
+                  "Unknown type of %s %s",
+                  PCollectionFusibility.class.getSimpleName(), fusibility));
+      }
+    }
+
+    return ImmutableExecutableStage.of(
+        environment,
+        inputPCollection,
+        fusedTransforms.build(),
+        ImmutableSet.copyOf(materializedPCollections));
+  }
+
+  private static Environment getStageEnvironment(
+      QueryablePipeline pipeline, Set<PTransformNode> initialNodes) {
+    IllegalArgumentException missingEnv =
+        new IllegalArgumentException(
+            String.format(
+                "%s must be populated on all %s in a %s",
+                Environment.class.getSimpleName(),
+                PTransformNode.class.getSimpleName(),
+                GreedyStageFuser.class.getSimpleName()));
```

Comment:
Don't construct the exception unconditionally. Since you're passing in a lambda anyway, you might just define a supplier here. Alternatively, refactor this into a separate method.

Reply:
Done.